src/poprox_recommender/components/joiners/fill.py

```diff
@@ -1,11 +1,15 @@
 from poprox_concepts import ArticleSet
 from poprox_recommender.lkpipeline import Component
+from poprox_recommender.lkpipeline.types import Lazy
 
 
 class Fill(Component):
     def __init__(self, num_slots):
         self.num_slots = num_slots
 
-    def __call__(self, candidates1: ArticleSet, candidates2: ArticleSet) -> ArticleSet:
-        articles = candidates1.articles + candidates2.articles
+    def __call__(self, candidates1: ArticleSet, candidates2: Lazy[ArticleSet]) -> ArticleSet:
+        articles = candidates1.articles
+        if len(articles) < self.num_slots:
+            articles = articles + candidates2.get().articles
```

Comment:
This is the key bit that this change enables for current POPROX components — `Fill`'s second-set upstream is only executed if the first branch fails to return enough nodes.